core/shared/src/main/scala/org/http4s/multipart/MultipartParser.scala

```diff
@@ -730,6 +732,168 @@ object MultipartParser {
       ) *> deferred.get.rethrow
     }
 
+  // //////////////////////////////////////
+  // Resource-safe receiver-based parser //
+  // //////////////////////////////////////
+
+  /** Pipe that receives PartStart / PartChunk / PartEnd events,
+    * extracting the chunk data between each PartStart / PartEnd
+    * pair and feeding it through a new `partDecoder` for each part.
+    * Emits the results from each `partDecoder`.
+    *
+    * Resources allocated by the returned Pipe are supervised by
+    * the given `supervisor`, such that when the `supervisor` closes,
+    * the allocated resources also close.
+    *
+    * The `partDecoder` may or may not actually *consume* the part body.
+    * The underlying parser will wait at a `PartEnd` event until the
+    * consumer finishes, i.e. the point when its Resource allocates a value
+    * or raises an error. If the Resource never allocates, the Pipe will
+    * hang forever.
+    *
+    * @param supervisor  A Supervisor used to restrict the lifetimes
+    *                    of resources allocated by this Pipe
+    * @param partDecoder A function that consumes the body of a part.
+    *                    Will be called once for each `PartStart` in
+    *                    the incoming stream.
+    * @tparam A Value type returned for each successfully-parsed Part
+    * @return A pipe that transforms Part events into corresponding
+    *         decoded values.
+    */
+  private[this] def decodePartEventsSupervised[F[_], A](
+      supervisor: Supervisor[F],
+      partDecoder: Part[F] => DecodeResult[Resource[F, *], A],
```

Comment:
`DecodeResult[Resource[F, *], A]` shouldn't this be inverted inside out in a sense of effect?

Reply:
I'll revisit this next week when I have some more time, and it's been a little while since my head was fully around all this, but I think it was important for `Resource` to be the outer type. That said, it might be possible to instead use `Part[F] => Resource[F, EntityDecoder[A, A]]` here.

Reply:
Ah, if I change to `partDecoder: Part[F] => Resource[DecodeResult[F, *], A]` then things get complicated...

---

With `DecodeResult[Resource[F, *], A]` I can just unwrap things with `.value` and have a `Resource[F, Either[DecodeFailure, A]]` which keeps me in `F` for the supervisor and the side-effects.

---

With `Resource[DecodeResult[F, *], A]` the side-effects need to be lifted into `EitherT` and the `useForever` call needs to be somehow converted from a `Fiber[DecodeResult, Nothing]` back down to a `Fiber[F, Nothing]`.

```scala
supervisor.supervise[Nothing] {
  partDecoder(Part(ps.value, channel.stream.unchunks)).value.attempt.evalTap { r =>
    resultPromise.complete(r.flatten) *> channel.close
  }.useForever.value.map {
    ??? // what to do with an Either[DecodeError, Nothing]
  }
}
```
---

With `Resource[F, EntityDecoder[F, A]]` it's a little better but the Part's body stream has to be constructed/referenced twice, which seems pitfall-y given that the body is coming from a Channel in this context:

```scala
supervisor.supervise[Nothing] {
  partDecoder(Part(ps.value, channel.stream.unchunks)).evalMap { decoder =>
   decoder.decode(Part(ps.value, channel.stream.unchunks), strict = false).value
  }.attempt.evalTap { r =>
    resultPromise.complete(r.flatten) *> channel.close
  }.useForever
}
```

This problem would be alleviated if I used `Headers` instead of `Part` as the argument to `partDecoder`, but the typical implementations want to inspect the name/filename values, which currently come from the `Part` class, so they'd have to copy Part's implementation logic in order to compensate.

---

I'm thinking the best option is to just leave this as-is.